weblate/formats/xwiki.py

```diff
@@ -0,0 +1,195 @@
+
+XML_HEADER = """<?xml version="1.1" encoding="UTF-8"?>
+
+<!--
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+-->
+
+"""
+
+
+class XWikiUnit(PropertiesUnit):
+    """
+        Inspired from PropertiesUnit, allow to override the methods to use the right
+        XWikiDialect methods for decoding properties.
+    """
+
+    @cached_property
+    def source(self):
+        # Need to decode property encoded string
+        return quote.xwiki_properties_decode(super().source)
+
+    @cached_property
+    def target(self):
+        """Return target string from a Translate Toolkit unit."""
+        if self.unit is None:
+            return ""
+        # Need to decode property encoded string
+        # This is basically stolen from
+        # translate.storage.properties.propunit.gettarget
+        # which for some reason does not return translation
+        value = quote.xwiki_properties_decode(self.unit.value)
+        value = re.sub("\\\\ ", " ", value)
+        return value
+
+
+class XWikiPropertiesFormat(PropertiesFormat):
+    """
+    Represents an XWiki Java Properties translation file.
+    This format specification is detailed in
+    https://dev.xwiki.org/xwiki/bin/view/Community/XWiki%20Translations%20Formats/#HXWikiJavaProperties
+    """
+
+    unit_class = XWikiUnit
+    name = 'XWiki Java Properties'
+    format_id = 'xwiki-java-properties'
+    loader = xwikifile
+    language_format = 'java'
+
+    def save_content(self, handle):
+        current_units = self.all_units
+        self.store.units = []
+        # Ensure that not translated units are saved too as missing properties.
+        for unit in current_units:
+            if unit.unit is None:
+                if not unit.has_content():
+                    unit.unit = unit.mainunit
+                else:
+                    missingunit, added = self.find_unit(unit.context, unit.source)
+                    unit.unit = missingunit.unit
+                    unit.unit.missing = True
+            self.add_unit(unit.unit)
+
+        self.store.serialize(handle)
+
+
+class XWikiPageProperties(xwikifile):
```

Comment:
Shouldn't this be part of the translate-toolkit instead?

Reply:
That could be yes. I guess in that case I should also move `XWikiFullPage`, but that certainly makes sense too.

Reply:
Basically anything file format related should go to the translate-toolkit, Weblate should contain only thin layer on top of that (ideally it would be really thin, but in reality the formats behave slightly differently, so more logic is needed).